### DISCUSSION TITLE
total_debt

### DISCUSSION THREAD

**Discussion by mrx23dot** [manusimidt/py-xbrl#36] [General]
Is there another lib that builds on this one to calc common values from GAAP, eg. total_debt etc?

**Comment by manusimidt** [comment]
I don't think so. I released this library in January of this year, so it's fairly new. I am not currently aware of any libraries that build on this.

**Reply by mrx23dot** [reply]
Or maybe formulas that use exact XBRL GAAP terms to calculate complex things like FreeCashFlow, long term debt etc.

Many websites have different formulas to calc these, and even then it's error prone to find the appropriate GAAP terms (without an business degree). Cheers

**Comment by mrx23dot** [comment]
Like 
Total debt = us-gaap:LongTermDebtCurrent + us-gaap:LongTermDebtNoncurrent + us-gaap:LinesOfCreditCurrent

**Reply by manusimidt** [reply]
@mrx23dot Sorry for the late reply, i was on vacation. 

The generalization of the calculation of consolidated financial items is very difficult because each company lists different items in its financial statements. 
For example, JPM lists the concept `us-gaap:FederalFundsPurchasedAndSecuritiesSoldUnderAgreementsToRepurchase` in their [balance sheet from 04.05.2021](https://www.sec.gov/ix?doc=/Archives/edgar/data/19617/000001961721000292/jpm-20210331.htm#ifecb6da2112c4adc964fef2d0c4d2285_181) (which, as far as I know, should also be included in total debt since it is a short term borrowing from another bank). 

It would probably be feasible to build another library on top of this one, which makes further calculations from existing us-gaap tags. However, this would require extensive accounting know-how, which I unfortunately do not have.

**Comment by mrx23dot** [comment]
Fare enough, I will have a look around what exists.

When I extract a fact which has dimensions, how do I access the date for the dimensions? eg
assets : [99, 98, 97]  -> [2020-06, 2019-06, 2018-06]

**Comment by mrx23dot** [comment]
Found the date in fact.context

Regarding dimensions, taking:
   https://www.sec.gov/ix?doc=/Archives/edgar/data/0000320193/000032019320000096/aapl-20200926.htm
   https://www.sec.gov/Archives/edgar/data/320193/000032019320000096/aapl-20200926_htm.xml

When I want to extract **Gross margin** 's latest value **24,689** (first column in pdf) this is what the lib gives back, how do I correlate the first item (24689) to the selected one?

Can't we sort it in the lib so first item would always be the latest?

2019-09-29 to 2020-09-26 0 dimension | GrossProfit: 104956000000.0 <--- not the first in html
2018-09-30 to 2019-09-28 0 dimension | GrossProfit: 98392000000.0
2017-10-01 to 2018-09-29 0 dimension | GrossProfit: 101839000000.0
2020-06-28 to 2020-09-26 0 dimension | GrossProfit: 24689000000.0 <---- this is the first in html
2020-03-29 to 2020-06-27 0 dimension | GrossProfit: 22680000000.0
2019-12-29 to 2020-03-28 0 dimension | GrossProfit: 22370000000.0
2019-09-29 to 2019-12-28 0 dimension | GrossProfit: 35217000000.0
2019-06-30 to 2019-09-28 0 dimension | GrossProfit: 24313000000.0
2019-03-31 to 2019-06-29 0 dimension | GrossProfit: 20227000000.0
2018-12-30 to 2019-03-30 0 dimension | GrossProfit: 21821000000.0
2018-09-30 to 2018-12-29 0 dimension | GrossProfit: 32031000000.0

**Reply by manusimidt** [reply]
I am not quite sure why you just want to extract the first fact that appears in the html code. 
Normally I would expect that you want to extract the fact for one specific timeframe because the arrangement of the facts in the filing is up to the filer.

However the libary uses ElementTree which normally preserves the order of the facts.
If you use the search feature of the ixbrl viewer of the SEC, you will see that the facts are displayed the same order as in the libary. 
![image](https://user-images.githubusercontent.com/29599104/123269351-2f491600-d4ff-11eb-97f2-4b78dcf90669.png)

**Reply by manusimidt** [reply]
Additionally, please note that you are not parsing the original filing. You are using a version that was converted by the SEC from iXBRL to XBRL (probably for compatibility reasons). If you want to make sure that the parsed version exactly matches the original filing you can also parse the original filing (iXBRL) of the company:

```python
cache: HttpCache = HttpCache('./cache')
xbrlParser = XbrlParser(cache)
# this url leads to the XBRL file that was converted by the SEC from the original iXBRL filing
# xbrl_url = 'https://www.sec.gov/Archives/edgar/data/320193/000032019320000096/aapl-20200926_htm.xml'
# the following url is the original filing url (iXBRL)
ixbrl_url = 'https://www.sec.gov/Archives/edgar/data/0000320193/000032019320000096/aapl-20200926.htm'
# the parser will automatically detect if the document uses the xbrl or ixbrl standard
inst: XbrlInstance = xbrlParser.parse_instance(ixbrl_url)
```

**Reply by mrx23dot** [reply]
Thank you, that makes sense, 
first entry is year period, 
later one is quarter period (which is more recent),

I just need to match the latest end date, then the shortest period.

  GrossProfit 104956000000.0 [] 2019-09-29 to 2020-09-26 0 dimension GrossProfit: 104956000000.0
  GrossProfit 24689000000.0 [] 2020-06-28 to 2020-09-26 0 dimension GrossProfit: 24689000000.0   <-- latest

I will also try out the ixbrl. The fun part begins when different companies do the filling slightly different way.

Cheers

**Comment by mrx23dot** [comment]
How come some documents don't have any metadata? (extractable numbers)
https://www.sec.gov/ix?doc=/Archives/edgar/data/0001403570/000149315221006582/form10-k.htm

What to do in these cases? Back to html greping?

**Reply by manusimidt** [reply]
The file you are referencing is no iXBRL document. Please use the XBRL Instance that is associated with this submission.
https://www.sec.gov/Archives/edgar/data/1403570/000149315221006582/qtmm-20190630.xml
You can see if a submission uses XBRL or iXBRL in the [index file](https://www.sec.gov/Archives/edgar/data/1403570/000149315221006582/0001493152-21-006582-index.html) of the submission.
![image](https://user-images.githubusercontent.com/29599104/123378407-01121780-d58d-11eb-97c6-ec93f1df4570.png)

**Reply by manusimidt** [reply]
If you want to distinguish programmatically between XBRL and iXBRL, I would recommend the [Structured Disclosure RSS Feeds](https://www.sec.gov/structureddata/rss-feeds-submitted-filings) from the SEC. 
![image](https://user-images.githubusercontent.com/29599104/123378829-88f82180-d58d-11eb-98a7-23514788c8e1.png)

**Comment by mrx23dot** [comment]
I created a higher level extractor building on this lib

https://github.com/mrx23dot/sec_extractor

still in early stage, but already usable.
Eventually it will have correct Free cash flow.

**Reply by manusimidt** [reply]
cool, i will check it out :)

**Comment by mrx23dot** [comment]
Found a possible parsing error:

https://www.sec.gov/ix?doc=/Archives/edgar/data/0000320193/000032019320000096/aapl-20200926.htm

**CommonStockParOrStatedValuePerShare**  should have 2 instances, but the lib only gives back 09/28/2019, even when not filtering for len(fact.context.segments)

Common Stock, Par or Stated Value Per Share
As of 09/26/2020
0.00001

Common Stock, Par or Stated Value Per Share
As of 09/28/2019
0.00001

Same problem for
- CommonStockSharesAuthorized  (missing As of 09/26/2020)
- EffectiveIncomeTaxRateReconciliationAtFederalStatutoryIncomeTaxRate (missing 12 months ending 09/28/2019)

**Reply by manusimidt** [reply]
Thank you for reporting! 
See #40. I will implement a fix in the next week or so :)
